### ISSUE TITLE
"certbot certificates" should check validity with OpenSSL

### ISSUE THREAD

**Issue by pde** [certbot/certbot#3866]
All sorts of weird things could have happened to the certs on disk to make them invalid, so it might be good for `certbot certificates` to run them through `openssl verify` or equivalent, and report any failures.

**Comment by pde** [comment]
Doing this with `python-cryptography` and `PyOpenSSL` would be theoretically desirable, if we can do it without breaking old distros versioned dependencies. Otherwise, we should just shell out.

**Comment by bmw** [comment]
@alex, @reaperhulk: is there a user facing way to do this with `PyOpenSSL<=0.13`? We've been unable to find one.

**Comment by reaperhulk** [comment]
I'm not familiar with a way to do this, but my knowledge of "old" pyOpenSSL isn't great. @hynek do you happen to know if there was a way?

**Comment by hynek** [comment]
No idea; I mean we’re talking 2011 here.  But it’s doubtful since it was a C-Ext back then so everything had to be consciously hard coded.

**Comment by ynasser** [comment]
If verification fails, do we want to still display the certificates as usual, but with an error message next to them, like this:

```
Found the following certs:
  Certificate Name: www.example.com
    Domains: www.example.com example.com
    Expiry Date: 2017-04-19 22:27:00+00:00 (VALID: 86 days)
    Certificate Path: /etc/letsencrypt/live/www.example.com/fullchain.pem
    Private Key Path: /etc/letsencrypt/live/www.example.com/privkey.pem
    Warning: OpenSSL verification for this cert failed!
```
... or should they just be handled in a similar fashion to cert parsing failures, and just be displayed at the end after the existing error message: `The following renewal configuration files were invalid: error1, error2, etc ... `.

**Comment by bmw** [comment]
OK thanks for the info. I wanted to verify to avoid shelling out to `openssl` if possible. We unfortunately have to support old versions of `PyOpenSSL` like `0.13.x` as it is the newest version available in some of the distros we are packaged such as EPEL on CentOS/RHEL 7.

@ynasser, as I'm sure you're aware, this means you'll have to shell out to `openssl` if you want to keep working on this issue. You should test against multiple versions of the `openssl` command line to make sure the command you're using works in different versions.

As for where to put the error message, you should do whatever you think looks best, but I'm leaning towards displaying the message at the end like we do for invalid renewal conf files. I suspect users are more likely to see the message if it is shown there.

**Comment by reaperhulk** [comment]
What kind of verification are you looking to do here? If you want to verify that the private key matches the public key in the cert or that the cert's signature matches the public key of the issuing CA those are both things that can be done via cryptography.

**Comment by bmw** [comment]
Really? That'd be great!

The main thing we're trying to do is verify that one cert's signature matches the public key in another certificate. Verifying the private key matches the public key in a cert might be worthwhile too though.

**Comment by pde** [comment]
Basically you can think of this as being "if I put this cert up on a server on the Internet, is it going to work?" I think the main things we're checking here are:

* the cert file/chain hasn't been corrupted, or accidentally overwritten by the sys admin with different certs, or whatever
* the cert and chain are actually signed by a CA that's trusted by browsers (this should always be the case for LE's ACME endpoint, but maybe not if people run other ACME servers)
* datetime range. Currently we check on the expiry end of the range. But certs can also be "not yet valid". It may be that that only ever happens because the user is running Certbot on a machine with an incorrect clock. So in that case we probably want to print an error message saying "Probably your clock is wrong, or the ACME server has done something very weird".
* any other weird properties of certs that we haven't thought of but which might be important. There are theoretically a lot of those (see ef [Table V of this paper](https://www.cs.cornell.edu/~shmat/shmat_oak14.pdf), though again it would be very surprising if an ACME server ever issued a cert that failed such tests.

I thiiiink we already effectively check that the cert is in its datetime validity window, but implicitly we'll also be doing this

**Comment by ynasser** [comment]
@pde: I was just going to use something like `openssl verify -CAfile chain.pem cert.pem`. The [verify](https://www.openssl.org/docs/man1.0.1/apps/verify.html) command seems to handle the main points you mentioned above, including checking that the cert is in the datetime validity window (see `X509_V_ERR_CERT_NOT_YET_VALID` and `X509_V_ERR_CERT_HAS_EXPIRED` in the verify documentation above). Does that seem reasonable? (Also, I've not really used openssl before, so I might be completely wrong about everything.)

@bmw: how far back should I go when it comes to testing this with multiple openssl versions?

**Comment by bmw** [comment]
@pde, what do you expect to happen for staging certs that aren't trusted by browsers?

> how far back should I go when it comes to testing this with multiple openssl versions?

1.0.0 should be the oldest version we support. I'd mainly just test on older distros like CentOS 6 and Debian 7 using `docker` or Digital Ocean droplets.

**Comment by ynasser** [comment]
The nice, short solution I had using `openssl` worked very nicely on Ubuntu 16.04, but not with older versions of `openssl` on CentOS 6.8.

Using `python-cryptography` instead to verify that the cert's signature matches the CA's public key seems fairly straightforward from the documentation, but I keep getting an `InvalidSignature` exception when I run this:

```
from cryptography import x509
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import padding
from cryptography.hazmat.backends import default_backend

cert_file = open('/etc/letsencrypt/live/www.example.com/cert.pem')
cert_data = cert_file.read()
cert = x509.load_pem_x509_certificate(data=cert_data, backend=default_backend())

# chain contains the Let's Encrypt certificate 
chain_file = open('/etc/letsencrypt/live/www.example.com/chain.pem')
chain_data = chain_file.read()
chain = x509.load_pem_x509_certificate(data=chain_data, backend=default_backend())

public_key = chain.public_key()

verifier = public_key.verifier(
                signature = cert.signature,
                padding = padding.PSS(
                              mgf = padding.MGF1(hashes.SHA256()),
                              salt_length = padding.PSS.MAX_LENGTH),
                algorithm = hashes.SHA256())
verifier.update(cert.tbs_certificate_bytes)
verifier.verify()
```

Does anyone have any idea as to what I'm doing wrong? The certificates themselves are fine, as https is working and `openssl verify -CAfile chain.pem cert.pem` checks out.

**Comment by bmw** [comment]
So I didn't dig deep into that code to see why it's not working, but this seems to work for me:
```
from cryptography import x509
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import padding
from OpenSSL.crypto import verify
from certbot import crypto_util

with open('chain.pem') as f:
    chain = crypto_util.pyopenssl_load_certificate(f.read())[0]

with open('cert.pem', 'rb') as f:
    cert = x509.load_pem_x509_certificate(f.read(), default_backend())

hash_name = cert.signature_hash_algorithm.name

verify(chain, cert.signature, cert.tbs_certificate_bytes, hash_name)
```

**Comment by reaperhulk** [comment]
@ynasser X509 certificates are currently signed via PKCS1v15 and not PSS, so that's one issue. The other one would be that it's only going to read the first cert from `chain.pem` so if that's not the cert that signed the end entity the verification will fail.

**Comment by ynasser** [comment]
@reaperhulk thank you for pointing out my mistakes! I'm still learning, so your feedback is very helpful.

@bmw thank you as well! I wasn't super familiar with `certbot/crypto_util`, and will make use of your example :)

**Comment by bmw** [comment]
Fixed in #4155.
